packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -1128,9 +1135,47 @@ class _CupertinoActionSheetActionState extends State<CupertinoActionSheetAction>
     widget.onPressed();
   }
 
+  // Calculates the font size for action sheet buttons, which deviate from
+  // standard HIG specifications.
+  //
+  // There is a non-linear relationship between the font size for regular body
+  // text and the font size for the action sheet buttons:
+  //
+  //  Text scale  | xs |  s |  m |  l | xl | xxl | xxxl | ax1 | ax2 | ax3 | ax4 | ax5
+  //  Body font   | 14 | 15 | 16 | 17 | 19 |  21 |  23  |  28 |  33 |  40 |  47 |  53
+  //  Button font | 21 | 21 | 21 | 21 | 23 |  24 |  24  |  28 |  33 |  40 |  47 |  53
+  //
+  // The `contextBodySize` is the body font size specified by context. The
+  // return value is the button font size, including the effect of context font
+  // scale factor. Divide by context font scale factor before using in a `Text`.
+  static double _buttonFontSize(double contextBodySize) {
+    // For very small or very large text, simpler scaling mappings are used.
+    if (contextBodySize <= 17) {
+      return 21;
+    }
+    if (contextBodySize >= 24.5) {
+      return contextBodySize;
+    }
+    // For mid-sized text, an interpolated curve is used. The breakpoint for
+    // switching between these behaviors is 24.5 instead of 28 to allow closer
+    // interpolation over a smaller range.
+    final double resultSize = -18.8 + 3.68 * contextBodySize * (1 - 0.02128 * contextBodySize);
```

Comment:
Seems like a lot of magic number. can you declare const in this method with appropriate name for readability?

Reply:
I don't know... because this curve is entirely interpolated. Let me ask around how people feel about it.

Reply:
can you explain a bit more how do you come up with this polynomial?

Reply:
Sure!
- I used the table above, and excluded the high end and the low end because they have simple rules, which leaves this:
```
  //  Text scale  |  l | xl | xxl | xxxl | ax1  
  //  Body font   | 17 | 19 |  21 |  23  |  28  
  //  Button font | 21 | 23 |  24 |  24  |  28  
```
- I cut the break point on the high end closer for better interpolation

```
  //  Text scale  |  l | xl | xxl | xxxl |   --  
  //  Body font   | 17 | 19 |  21 |  23  |  24.5
  //  Button font | 21 | 23 |  24 |  24  |  24.5  
```
- I sent this table to a cubit interpolation and got the result formula.

Reply:
I see, yeah In that case, there is probably not much to explain in the code.

It would be good to mention this polynomial is generated though interpolation tool.

Reply:
That's a good idea. I'll probably just record the method above in the doc.

Reply:
curious is the output not normalized as ax^2 + bx + c? The original format makes me think this is hand calculated, that is why I was wondering if you can define const to improve readability

Reply:
I plotted the original interpolation function on Desmos; unfortunately it doesn't line up exactly.

![desmos old](https://github.com/flutter/flutter/assets/10457200/6a3207a2-d8d0-4c34-97f7-a34ba969c074)

<br>

I discovered that the function $y=\frac{1}{16}(x-17)^2+21$ happens to match the slope of both lines exactly!

![desmos new](https://github.com/flutter/flutter/assets/10457200/4c696536-018f-45a1-b1e2-d9b89210a019)

<br>

Here's the function, translated to Dart code:

```dart
double _buttonFontSize(double contextBodySize) {
  const double minSize = 21.0;
  const double lowerBound = 17.0;
  const double upperBound = 25.0;

  if (contextBodySize <= lowerBound) {
      return minSize;
    }
  if (contextBodySize >= upperBound) {
    return contextBodySize;
  }

  final double x = contextBodySize - lowerBound;

  // quadratic interpolation
  return minSize + x * x / 16;
}
```

Reply:
@nate-thegrate Yes, the ends don't align exactly, but since there's a `roundToDouble` at the end the fractional part doesn't matter.

For your second curve, they don't match the table that I derived (as noted in the doc).

Reply:
> curious is the output not normalized as ax^2 + bx + c? 

The current form saves a multiplication... I mean, yes it's not much, but it's free. :/

Reply:
@nate-thegrate I updated the doc and used the variable names in your doc! Thanks for the suggestion!